cmd/state/generate/generate_header_snapshot.go

```diff
@@ -0,0 +1,114 @@
+package generate
+
+import (
+	"errors"
+	"fmt"
+	"github.com/anacrolix/torrent/bencode"
+	"github.com/anacrolix/torrent/metainfo"
+	"github.com/ledgerwatch/turbo-geth/common"
+	"github.com/ledgerwatch/turbo-geth/common/dbutils"
+	"github.com/ledgerwatch/turbo-geth/core/rawdb"
+	"github.com/ledgerwatch/turbo-geth/ethdb"
+	"github.com/ledgerwatch/turbo-geth/log"
+	"github.com/ledgerwatch/turbo-geth/turbo/torrent"
+	"math/big"
+	"os"
+	"os/signal"
+	"time"
+)
+
+func HeaderSnapshot(dbPath, snapshotPath string, toBlock uint64) error {
+	if snapshotPath == "" {
+		return errors.New("empty snapshot path")
+	}
+	err := os.RemoveAll(snapshotPath)
+	if err != nil {
+		return err
+	}
+	kv := ethdb.NewLMDB().Path(dbPath).MustOpen()
+	snkv := ethdb.NewLMDB().WithBucketsConfig(func(defaultBuckets dbutils.BucketsCfg) dbutils.BucketsCfg {
+		return dbutils.BucketsCfg{
+			dbutils.HeaderPrefix:       dbutils.BucketConfigItem{},
+			dbutils.SnapshotInfoBucket: dbutils.BucketConfigItem{},
+			dbutils.HeadBlockKey:       dbutils.BucketConfigItem{},
+		}
+	}).Path(snapshotPath).MustOpen()
+
+	db := ethdb.NewObjectDatabase(kv)
+	sndb := ethdb.NewObjectDatabase(snkv)
+
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt)
+	t := time.Now()
+	chunkFile := 30000
+	tuples := make(ethdb.MultiPutTuples, 0, chunkFile*3)
+	var hash common.Hash
+	var header []byte
+	for i := uint64(1); i <= toBlock; i++ {
+		select {
+		case <-ch:
+			return errors.New("interrupted")
+		default:
+
+		}
+		hash = rawdb.ReadCanonicalHash(db, i)
+		header = rawdb.ReadHeaderRLP(db, hash, i)
+		tuples = append(tuples, []byte(dbutils.HeaderPrefix), dbutils.HeaderKey(i, hash), header)
```

Comment:
So the header snapshot contains one out of three types of records that are in the `HeaderPrefix` table, right?

Reply:
Yes, you are right. Just list of canonical headers. We can build header number to header hash/td and canonical indexes after snapshot download.